libpromises/evalfunction.c

```diff
@@ -6659,6 +6659,158 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
+    char *line;
+    size_t line_len = 0;
+
+    // Current line number, for debugging
+    size_t line_number = 0;
+
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
+    {
+        Seq *seq = SeqParseCsvString(line);
+        ++line_number;
+        free(line);
+        if (seq == NULL)
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Failed to parse line %lu, line ignored.",
+                fp->name, line_number);
+            continue;
+        }
+
+        if (line_len == 0)
+        {
+            line_len = SeqLength(seq);
+        }
+        else if (line_len != SeqLength(seq))
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Line %lu has incorrect amount of elements, "
+                "%lu instead of %lu. Line ignored.",
+                fp->name, line_number, SeqLength(seq), line_len);
+            SeqDestroy(seq);
+            continue;
+        }
+
+        if (has_heading && heading_seq == NULL)
+        {
+            Log(LOG_LEVEL_DEBUG,
+                "%s: Found header at line %lu",
+                fp->name, line_number);
+            SeqRemove(seq, col);
+            heading_seq = seq;
+        }
+        else
+        {
+            if (!IsDefinedClass(ctx, SeqAt(seq, col)))
+            {
+                SeqDestroy(seq);
+                continue;
+            }
+
+            SeqRemove(seq, col);
+            size_t len = SeqLength(seq);
+
+            JsonElement *json_object = JsonObjectCreate(len);
+            for (size_t i = 0; i < len; i++)
+            {
+                if (has_heading)
+                {
+                    JsonObjectAppendString(json_object,
+                                           SeqAt(heading_seq, i),
+                                           SeqAt(seq, i));
+                }
+                else
+                {
+                    size_t key_len = PRINTSIZE(size_t);
+                    char key[key_len];
+                    xsnprintf(key, key_len, "%lu", i);
+
+                    JsonObjectAppendString(json_object,
+                                           key,
+                                           SeqAt(seq, i));
+                }
+            }
+
+            JsonArrayAppendObject(json, json_object);
+            SeqDestroy(seq);
+        }
+    }
+
+    if (sort_col == col)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "%s: sorting column (%lu) is the same as class expression "
+            "column (%lu). Not sorting data container.",
+            fp->name, sort_col, col);
+    }
+    else if (sort_col > line_len)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "%s: sorting index %lu out of bounds. Not sorting data "
+            "container.",
+            fp->name, sort_col);
+    }
+    else
+    {
+        if (sort_col > col)
+        {
+            sort_col--;
+        }
+
+        JsonSort(json, ContainerComparator, &sort_col);
+    }
+
+    fclose(csv_file);
+    if (heading_seq != NULL)
```

Comment:
Not a big deal, but I would also rename `heading_seq`. I don't think you need to specify type info in the name (except if there were 2 variables, `heading` and `heading_seq`), so I would prefer just a description of what it is, e.g. `heading`, `column_labels`, `column_headings`, or something similar.